File: packages/node/src/client.test.ts
import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";
import { WalandClient } from "./client.js";
import { WalandError, WalandValidationError } from "./errors.js";

const API_KEY = "waland_test_key";
const SESSION_ID = "session-abc123";

const successBody = {
  id: "log-id",
  sessionId: SESSION_ID,
  organizationId: "org-id",
  chatId: "8801712345678@s.whatsapp.net",
  text: "Hello",
  mediaUrl: null,
  status: "sent" as const,
  messageId: "wa-msg-id",
  error: null,
  createdAt: "2026-05-24T10:00:00.000Z",
};

const checkNumberBody = {
  number: "8801712345678",
  chatId: "8801712345678@s.whatsapp.net",
  jid: "8801712345678@s.whatsapp.net",
  exists: true,
};

describe("WalandClient", () => {
  beforeEach(() => {
    vi.stubGlobal(
      "fetch",
      vi.fn().mockResolvedValue(
        new Response(JSON.stringify(successBody), {
          status: 201,
          headers: { "Content-Type": "application/json" },
        }),
      ),
    );
  });

  afterEach(() => {
    vi.unstubAllGlobals();
  });

  it("requires apiKey and sessionId", () => {
    expect(() => new WalandClient("", SESSION_ID)).toThrow(
      WalandValidationError,
    );
    expect(() => new WalandClient(API_KEY, "")).toThrow(WalandValidationError);
  });

  it("sends a text message with bearer auth", async () => {
    const client = new WalandClient(API_KEY, SESSION_ID);

    const result = await client.sendMessage({
      chatId: "8801712345678@s.whatsapp.net",
      text: "Hello",
    });

    expect(result).toEqual(successBody);
    expect(fetch).toHaveBeenCalledOnce();
    expect(fetch).toHaveBeenCalledWith(
      "https://api.waland.dev/v1/sessions/session-abc123/send",
      expect.objectContaining({
        method: "POST",
        headers: expect.objectContaining({
          Authorization: `Bearer ${API_KEY}`,
          "Content-Type": "application/json",
        }),
        body: JSON.stringify({
          chatId: "8801712345678@s.whatsapp.net",
          text: "Hello",
        }),
      }),
    );
  });

  it("sends media with optional filename", async () => {
    const client = new WalandClient(API_KEY, SESSION_ID);

    await client.sendMessage({
      chatId: "8801712345678@s.whatsapp.net",
      text: "Caption",
      mediaUrl: "https://example.com/photo.jpg",
      mediaFilename: "photo.jpg",
    });

    expect(fetch).toHaveBeenCalledWith(
      expect.any(String),
      expect.objectContaining({
        body: JSON.stringify({
          chatId: "8801712345678@s.whatsapp.net",
          text: "Caption",
          mediaUrl: "https://example.com/photo.jpg",
          mediaFilename: "photo.jpg",
        }),
      }),
    );
  });

  it("rejects invalid chatId", async () => {
    const client = new WalandClient(API_KEY, SESSION_ID);

    await expect(
      client.sendMessage({ chatId: "not-a-jid", text: "Hi" }),
    ).rejects.toThrow(WalandValidationError);
  });

  it("rejects when neither text nor mediaUrl is provided", async () => {
    const client = new WalandClient(API_KEY, SESSION_ID);

    await expect(
      client.sendMessage({ chatId: "8801712345678@s.whatsapp.net" }),
    ).rejects.toThrow(WalandValidationError);
  });

  it("throws WalandError on API errors", async () => {
    vi.mocked(fetch).mockResolvedValueOnce(
      new Response(
        JSON.stringify({
          statusCode: 401,
          message: "Invalid or missing org API key",
          error: "Unauthorized",
        }),
        { status: 401, headers: { "Content-Type": "application/json" } },
      ),
    );

    const client = new WalandClient(API_KEY, SESSION_ID);

    await expect(
      client.sendMessage({
        chatId: "8801712345678@s.whatsapp.net",
        text: "Hi",
      }),
    ).rejects.toMatchObject({
      name: "WalandError",
      statusCode: 401,
      message: "Invalid or missing org API key",
    } satisfies Partial<WalandError>);
  });

  it("checks a number", async () => {
    vi.mocked(fetch).mockResolvedValueOnce(
      new Response(JSON.stringify(checkNumberBody), {
        status: 200,
        headers: { "Content-Type": "application/json" },
      }),
    );

    const client = new WalandClient(API_KEY, SESSION_ID);
    const result = await client.checkNumber({ number: "8801712345678" });

    expect(result).toEqual(checkNumberBody);
    expect(fetch).toHaveBeenCalledWith(
      "https://api.waland.dev/v1/sessions/session-abc123/check-number",
      expect.objectContaining({
        method: "POST",
        body: JSON.stringify({ number: "8801712345678" }),
      }),
    );
  });

  it("rejects empty number for checkNumber", async () => {
    const client = new WalandClient(API_KEY, SESSION_ID);

    await expect(client.checkNumber({ number: "  " })).rejects.toThrow(
      WalandValidationError,
    );
  });
});

File: packages/go/client_test.go
package waland

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"testing"
)

const (
	testAPIKey    = "waland_test_key"
	testSessionID = "session-abc123"
)

func TestNewClientRequiresCredentials(t *testing.T) {
	if _, err := NewClient("", testSessionID, nil); err == nil {
		t.Fatalf("expected error for empty api key")
	}

	if _, err := NewClient(testAPIKey, "", nil); err == nil {
		t.Fatalf("expected error for empty session ID")
	}
}

func TestSendTextMessage(t *testing.T) {
	t.Parallel()

	var gotAuth string
	var gotBody map[string]interface{}

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/sessions/session-abc123/send" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}

		gotAuth = r.Header.Get("Authorization")
		body, _ := io.ReadAll(r.Body)
		_ = json.Unmarshal(body, &gotBody)

		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusCreated)
		_, _ = w.Write([]byte(`{"id":"log-id","sessionId":"session-abc123","organizationId":"org-id","chatId":"8801712345678@s.whatsapp.net","text":"Hello","mediaUrl":null,"status":"sent","messageId":"wa-msg-id","error":null,"createdAt":"2026-05-24T10:00:00.000Z"}`))
	}))
	defer server.Close()

	client, err := NewClient(testAPIKey, testSessionID, &ClientOptions{BaseURL: server.URL})
	if err != nil {
		t.Fatalf("NewClient failed: %v", err)
	}

	result, err := client.SendMessage(context.Background(), SendMessageParams{
		ChatID: "8801712345678@s.whatsapp.net",
		Text:   "Hello",
	})
	if err != nil {
		t.Fatalf("SendMessage failed: %v", err)
	}

	if gotAuth != "Bearer "+testAPIKey {
		t.Fatalf("unexpected auth header: %s", gotAuth)
	}
	if gotBody["chatId"] != "8801712345678@s.whatsapp.net" || gotBody["text"] != "Hello" {
		t.Fatalf("unexpected request body: %#v", gotBody)
	}
	if result.Status != SmsLogStatusSent {
		t.Fatalf("unexpected status: %s", result.Status)
	}
}

func TestSendMediaMessage(t *testing.T) {
	t.Parallel()

	var gotBody map[string]interface{}

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		body, _ := io.ReadAll(r.Body)
		_ = json.Unmarshal(body, &gotBody)

		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusCreated)
		_, _ = w.Write([]byte(`{"id":"log-id","sessionId":"session-abc123","organizationId":"org-id","chatId":"8801712345678@s.whatsapp.net","text":"Caption","mediaUrl":"https://example.com/photo.jpg","status":"sent","messageId":"wa-msg-id","error":null,"createdAt":"2026-05-24T10:00:00.000Z"}`))
	}))
	defer server.Close()

	client, _ := NewClient(testAPIKey, testSessionID, &ClientOptions{BaseURL: server.URL})
	_, err := client.SendMessage(context.Background(), SendMessageParams{
		ChatID:        "8801712345678@s.whatsapp.net",
		Text:          "Caption",
		MediaURL:      "https://example.com/photo.jpg",
		MediaFilename: "photo.jpg",
	})
	if err != nil {
		t.Fatalf("SendMessage failed: %v", err)
	}

	if gotBody["mediaUrl"] != "https://example.com/photo.jpg" {
		t.Fatalf("missing mediaUrl: %#v", gotBody)
	}
	if gotBody["mediaFilename"] != "photo.jpg" {
		t.Fatalf("missing mediaFilename: %#v", gotBody)
	}
}

func TestRejectsInvalidChatID(t *testing.T) {
	client, _ := NewClient(testAPIKey, testSessionID, nil)

	_, err := client.SendMessage(context.Background(), SendMessageParams{
		ChatID: "not-a-jid",
		Text:   "Hi",
	})
	if err == nil {
		t.Fatalf("expected validation error")
	}
	if _, ok := err.(*ValidationError); !ok {
		t.Fatalf("expected ValidationError, got %T", err)
	}
}

func TestRejectsMissingTextAndMedia(t *testing.T) {
	client, _ := NewClient(testAPIKey, testSessionID, nil)

	_, err := client.SendMessage(context.Background(), SendMessageParams{
		ChatID: "8801712345678@s.whatsapp.net",
	})
	if err == nil {
		t.Fatalf("expected validation error")
	}
}

func TestReturnsAPIError(t *testing.T) {
	t.Parallel()

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusUnauthorized)
		_, _ = w.Write([]byte(`{"statusCode":401,"message":"Invalid or missing org API key","error":"Unauthorized"}`))
	}))
	defer server.Close()

	client, _ := NewClient(testAPIKey, testSessionID, &ClientOptions{BaseURL: server.URL})
	_, err := client.SendMessage(context.Background(), SendMessageParams{
		ChatID: "8801712345678@s.whatsapp.net",
		Text:   "Hi",
	})
	if err == nil {
		t.Fatalf("expected API error")
	}

	apiErr, ok := err.(*APIError)
	if !ok {
		t.Fatalf("expected APIError, got %T", err)
	}
	if apiErr.StatusCode != 401 {
		t.Fatalf("unexpected status code: %d", apiErr.StatusCode)
	}
	if apiErr.Message != "Invalid or missing org API key" {
		t.Fatalf("unexpected message: %s", apiErr.Message)
	}
}

func TestCheckNumber(t *testing.T) {
	t.Parallel()

	var gotBody map[string]interface{}

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/sessions/session-abc123/check-number" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}

		body, _ := io.ReadAll(r.Body)
		_ = json.Unmarshal(body, &gotBody)

		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		_, _ = w.Write([]byte(`{"number":"8801712345678","chatId":"8801712345678@s.whatsapp.net","jid":"8801712345678@s.whatsapp.net","exists":true}`))
	}))
	defer server.Close()

	client, _ := NewClient(testAPIKey, testSessionID, &ClientOptions{BaseURL: server.URL})
	result, err := client.CheckNumber(context.Background(), CheckNumberParams{Number: "8801712345678"})
	if err != nil {
		t.Fatalf("CheckNumber failed: %v", err)
	}

	if gotBody["number"] != "8801712345678" {
		t.Fatalf("unexpected request body: %#v", gotBody)
	}
	if result.Exists == nil || !*result.Exists {
		t.Fatalf("expected exists=true, got %#v", result.Exists)
	}
}

func TestCheckNumberRequiresNumber(t *testing.T) {
	client, _ := NewClient(testAPIKey, testSessionID, nil)

	_, err := client.CheckNumber(context.Background(), CheckNumberParams{Number: " "})
	if err == nil {
		t.Fatalf("expected validation error")
	}
	if _, ok := err.(*ValidationError); !ok {
		t.Fatalf("expected ValidationError, got %T", err)
	}
}
